netebpfext/net_ebpf_ext_tracelog.c

```diff
@@ -44,3 +44,500 @@ net_ebpf_ext_trace_terminate()
+#define CASE_XDP case _NET_EBPF_EXT_TRACELOG_KEYWORD_XDP
+
+#define LEVEL_LOG_ALWAYS NET_EBPF_EXT_TRACELOG_LEVEL_LOG_ALWAYS
+#define LEVEL_CRITICAL NET_EBPF_EXT_TRACELOG_LEVEL_CRITICAL
+#define LEVEL_ERROR NET_EBPF_EXT_TRACELOG_LEVEL_ERROR
+#define LEVEL_WARNING NET_EBPF_EXT_TRACELOG_LEVEL_WARNING
+#define LEVEL_INFO NET_EBPF_EXT_TRACELOG_LEVEL_INFO
+#define LEVEL_VERBOSE NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE
+
+#define CASE_LOG_ALWAYS case _NET_EBPF_EXT_TRACELOG_LEVEL_LOG_ALWAYS
+#define CASE_CRITICAL case _NET_EBPF_EXT_TRACELOG_LEVEL_CRITICAL
+#define CASE_LEVEL_ERROR case _NET_EBPF_EXT_TRACELOG_LEVEL_ERROR
+#define CASE_WARNING case _NET_EBPF_EXT_TRACELOG_LEVEL_WARNING
+#define CASE_INFO case _NET_EBPF_EXT_TRACELOG_LEVEL_INFO
+#define CASE_VERBOSE case _NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE
+
+#define NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE_KEYWORD_SWITCH(api_name, status)       \
+    switch (keyword) {                                                               \
+    CASE_BASE:                                                                       \
+        _NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(KEYWORD_BASE, api_name, status);      \
+        break;                                                                       \
+    CASE_EXT:                                                                        \
+        _NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(KEYWORD_EXT, api_name, status);       \
+        break;                                                                       \
+    CASE_BIND:                                                                       \
+        _NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(KEYWORD_BIND, api_name, status);      \
+        break;                                                                       \
+    CASE_SOCK_ADDR:                                                                  \
+        _NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(KEYWORD_SOCK_ADDR, api_name, status); \
+        break;                                                                       \
+    CASE_SOCK_OPS:                                                                   \
+        _NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(KEYWORD_SOCK_OPS, api_name, status);  \
+        break;                                                                       \
+    CASE_XDP:                                                                        \
+        _NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE(KEYWORD_XDP, api_name, status);       \
+        break;                                                                       \
+    default:                                                                         \
+        break;                                                                       \
+    }
+
+__declspec(noinline) void net_ebpf_ext_log_ntstatus_api_failure(
+    net_ebpf_ext_tracelog_keyword_t keyword, _In_z_ const char* api_name, NTSTATUS status)
+{
+    NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE_KEYWORD_SWITCH(api_name, status);
```

Comment:
Curious why `keyword` was _not_ passed as an argument (as well) to the `NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE_KEYWORD_SWITCH` macro?

Reply:
This is discussed in the overview section.